Ensure post_id is sent to the REST API in ServerSideRender component

alexsanford opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Fixes https://github.com/WordPress/gutenberg/issues/8178

This PR adds a `post_id` parameter to the REST API call of the `ServerSideRender` component. This ensures the global `$post` object is set properly. Previously, it was only set when rendering to the frontend, but not when rendering for the API call.

While working on this, we decided that this component, being tied closely to the WordPress REST API, really belongs in the `@wordpress/editor` package. It was moved in this PR, and a deprecated version was added in the original package (`@wordpress/components`) for backwards compatibility.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

This was tested locally using a dynamic block that displayed information from the `$post` object, ensuring the post information was rendered properly on the frontend and in the editor using `ServerSideRender`.

An automated test was added to ensure that the parameter is added properly to the URL path.

Also I verified that the current two blocks using `ServerSideRender` (Archives and Latest Comments) still work as expected.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

This change should be a non-breaking enhancement.

That said, I did change the name of the `rendererPathWithAttributes` function to `rendererPath`, and it's technically possible that some 3rd-party code is importing and using that. But I assume that's ok since it's not part of Gutenberg's documented API.

This change also deprecates `wp.components.ServerSideRender` in favor of `wp.editor.ServerSideRender`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

alexsanford committed:
Ensure post_id is sent to the REST API in ServerSideRender component

jorgefilipecosta reviewed (commented):
Hi @alexsanford, thank you for working on this issue 👍 I feel we should not use @wordpress/data in @wordpress/components (until now we never did it).
I think a solution could be moving the component to @worpress/editor (with a straightforward deprecation logic on the existing component) or creating a wrapper in @wordpress/editor that injects the post id using withSelect similarly to what happened with editor/utils/media-upload.

alexsanford committed:
Replace custom wrapper with withSelect HOC

alexsanford commented:
@jorgefilipecosta Thanks for your comments!

I think that moving this component into `@wordpress/editor` makes a lot of sense given that it's _not_ a generic UI component and it requires the WP API. I'll get that moved over 🙂

alexsanford committed:
Move ServerSideRender component to editor package

Since it accesses the WordPress API, it is not a pure UI component, and
should not be in the components package.

alexsanford committed:
Move ServerSideRender component into @wordpress/editor package

alexsanford committed:
Update core blocks to use ServerSideRender from @wordpress/editor

alexsanford commented on packages/components/src/server-side-render/index.js:
```diff
@@ -13,19 +13,21 @@ import {
 import { __, sprintf } from '@wordpress/i18n';
 import apiFetch from '@wordpress/api-fetch';
 import httpBuildQuery from 'http-build-query';
+import { select } from '@wordpress/data';
```
@jorgefilipecosta Done! I moved the component to `@wordpress/editor` and created a deprecated version in `@wordpress/components`.

I also updated the PR description to reflect this.

alexsanford commented on packages/components/src/server-side-render/index.js:
```diff
@@ -96,4 +98,24 @@ export class ServerSideRender extends Component {
 
-export default ServerSideRender;
+/**
+ * A wrapper to inject the Post Id.
```
Done

alexsanford commented:
@jorgefilipecosta I addressed the comments above. Note that I also rebased the branch.

danielbachhuber commented:
@alexsanford Thanks for taking this on :)

Per my note in the original issue, I think we want to go in a different direction:

> Similar to `editor/utils/media-upload`, we'll want to create a `editor/utils/server-side-render/` wrapper that enriches `ServerSideRender` with `getCurrentPostId()`.

Doing so lets us keep `ServerSideRender` as a generic component, but then also lets us have an enriched component in the editor context.

alexsanford commented:
@danielbachhuber but would you say that `ServerSideRender` really is a generic component? I suppose it is in the sense that it could be used outside of the editor. But it seems pretty closely tied to the WordPress API, which is unlike all of the other truly generic components.

In any case, I can put together another PR and we can decide which one we like better 🙂

danielbachhuber commented:
> would you say that `ServerSideRender` really is a generic component? I suppose it is in the sense that it could be used outside of the editor. But it seems pretty closely tied to the WordPress API, which is unlike all of the other truly generic components.

Yes, generic in the sense that it can be used outside of the editor. For instance, it will need to be used to preview blocks within the Customize experience (whenever that comes). Your wording is more correct, however.

> In any case, I can put together another PR and we can decide which one we like better 

Sounds good, thanks.

aduth reviewed (changes requested):
I was going to leave a comment about this not being generic component with its consideration of the post ID from editor state, but @danielbachhuber summarized much better than I already.

alexsanford commented:
@danielbachhuber Sorry for the delay on this one! I've submitted a new PR (https://github.com/WordPress/gutenberg/pull/9323) so let's continue the conversation over there 🙂 

Closing this one.
